src/scripts/hooks/ready.ts

```diff
@@ -134,15 +134,29 @@ export const Ready = {
             });
 
             // Now that all game data is available, Determine what actors we need to reprepare.
-            // Add actors currently in an encounter, then in a party, then all familiars, then parties
+            // Add actors currently in an encounter, then in a party, then all familiars, then parties, then in terrains
+            const inTerrains: ActorPF2e[] = [];
+            const hasSceneTerrains = !!canvas.scene?.flags.pf2e.terrainTypes?.length;
+            for (const token of canvas.scene?.tokens ?? []) {
+                if (!token.actor) continue;
+                if (hasSceneTerrains) {
+                    inTerrains.push(token.actor);
+                } else if ((token.regions ?? []).some((r) => r.behaviors.some((b) => b.type === "pf2eTerrain"))) {
+                    inTerrains.push(token.actor);
+                }
+            }
             const parties = game.actors.filter((a): a is PartyPF2e<null> => a.isOfType("party"));
-            const actorsToReprepare = R.compact([
-                ...game.combats.contents.flatMap((e) => e.combatants.contents).map((c) => c.actor),
-                ...parties.flatMap((p) => p.members).filter((a) => !a.isOfType("familiar")),
-                ...game.actors.filter((a) => a.type === "familiar"),
-                ...parties,
-            ]);
-            resetActors(new Set(actorsToReprepare), { sheets: false });
+            const actorsToReprepare = R.filter(
+                [
+                    ...game.combats.contents.flatMap((e) => e.combatants.contents).map((c) => c.actor),
+                    ...parties.flatMap((p) => p.members).filter((a) => !a.isOfType("familiar")),
+                    ...game.actors.filter((a) => a.type === "familiar"),
+                    ...parties,
+                    ...inTerrains,
```

Comment:
This should be towards the start. Its in this order for a reason: familiars depend on masters, and parties depend on members.

Reply:
Well, actually, depends on how Set handles it, now that I think about it.

Reply:
Maybe it's not the best idea to do that here then. Some of those actors could also be `inTerrains`. Leaving it in the `canvasReady` hook has a big chance of double preparations on first load.